doc/source/serve/dev-workflow.md

```diff
@@ -0,0 +1,123 @@
+
+- You can put all of your code in a single Python file.
+- You can put your call graph in a Python file and your deployment class definitions in other files.
+
+## Local Development using Python handles
+
+You can use `serve.run` in process to run and test your application locally, using a Python handle to send requests rather than HTTP.
+
+Benefits:
+
+- Self-contained Python is convenient for writing local integration tests.
+- No need to deploy to a cloud provider or manage infrastructure.
+
+Drawbacks:
+
+- Doesn't test HTTP endpoints.
+- Doesn't test multi-node clusters, can't use GPUs if your local machine doesn't have them.
+
+Let's see a simple example.
+
+```python
+# Filename: local_dev.py
+from ray import serve
+from ray.dag.input_node import InputNode
+import starlette.requests
+
+@serve.deployment
+def double(s: str):
+    return s + " " + s
+
+@serve.deployment
+class HelloDeployment:
+    def __call__(self, request: starlette.requests.Request):
+        return "Hello, {}!".format(request.params["name"])
+
+# Everything beyond this line could be in a different file, provided
+# HelloDeployment and double are imported.
+hello_deployment = HelloDeployment.bind()
+with InputNode() as http_request:
+    hello_output = hello_deployment.bind(http_request)
+    double_output = double.bind(hello_output)
+
+ref = double_output.execute("Ray")
+result = ray.get(ref)
+assert result = "Hello, Ray! Hello, Ray!"
+```
+
+Here we used the `execute` method to directly send a Python object to the deployment, rather than an HTTP request.
+
+% TODO: We didn't use serve.run() like we said at the begining; should we?
+
+
+## Local Development with HTTP requests
+
+You can use the `serve run` CLI command to run and test your application locally using HTTP to send requests (similar to how you might use `uvicorn run` if you're familiar with Uvicorn):
+
+```bash
+serve run local_dev:HelloDeployment
+```
+
+Now that Serve is running, we can pass in HTTP requests to the application and see the output.  For simplicity, we'll just use the `curl` command to send requests from another terminal.
+
+```bash
+curl -X GET http://localhost:8000/hello?name=Ray
+```
+
+
+
+## 3. Testing on a remote cluster
+
+When making the transition from your local machine to a remote cluster, you'll need to make sure your cluster has a similar environment to your local machine--files, environment variables, and Python packages, for example.  During development, you can use [Runtime Environments](runtime-environments.md) to manage this in a flexible way.
+
+See Ray Client (TODO: link) for more information on the Ray address specified here by the `--address` parameter.
+
+```bash
+serve run  --address=ray://<cluster-ip-address>:10001 --runtime-env-json='{"env_vars": {"MY_ENV_VAR": "my-value"}, "working_dir": "./project/src", "pip": ["requests", "chess"]}' local_dev:HelloDeployment
```

Comment:
can we add a `note` what ./project/src should usually include? Or Add a link for `working_dir`